File: Trees_Graphs/go/check_subtree.go
package main

import (
  "fmt"
  "types"
)


func isSubTree(node *types.TreeNodeWithParent, root2 *types.TreeNodeWithParent) bool {

  if node == root2 {
    return true
  }

  if node.LeftChild == nil && node.RightChild == nil {
    return false
  } else if node.LeftChild == nil {
    return isSubTree(node.RightChild, root2)
  } else if node.RightChild == nil {
    return isSubTree(node.LeftChild, root2)
  } else {
    return isSubTree(node.LeftChild, root2) || isSubTree(node.RightChild, root2)
  }

  return false
}


func main() {
  root := types.TreeNodeWithParent{Value: 4}

  secondLevelLeftChild := types.TreeNodeWithParent{Value: 2, Parent: &root}
  secondLevelRightChild := types.TreeNodeWithParent{Value: 6, Parent: &root}
  root.LeftChild = &secondLevelLeftChild
  root.RightChild = &secondLevelRightChild

  thirdLevelFirst := types.TreeNodeWithParent{Value: 1, Parent: &secondLevelLeftChild}
  thirdLevelSecond := types.TreeNodeWithParent{Value: 3, Parent: &secondLevelLeftChild}
  thirdLevelThird := types.TreeNodeWithParent{Value: 5, Parent: &secondLevelRightChild}
  thirdLevelFourth := types.TreeNodeWithParent{Value: 7, Parent: &secondLevelRightChild}

  secondLevelLeftChild.LeftChild = &thirdLevelFirst
  secondLevelLeftChild.RightChild = &thirdLevelSecond
  secondLevelRightChild.LeftChild = &thirdLevelThird
  secondLevelRightChild.RightChild = &thirdLevelFourth

  anotherTree := &secondLevelLeftChild

  fmt.Println(isSubTree(&root, anotherTree))
}

File: Trees_Graphs/python/check_subtree.py
from Tree import Tree


def is_subtree(node1, node2):
    if node1 == node2:
        return True

    if node1.left_child is None and node1.right_child is None:
        return False
    elif node1.left_child is None:
        return is_subtree(node1.right_child, node2)
    elif node1.right_child is None:
        return is_subtree(node1.left_child, node2)
    else:
        return is_subtree(node1.left_child, node2) or is_subtree(node1.right_child, node2)

    return False


def main():
    array = list(range(7))
    tree = Tree(array)
    tree.print_tree()

    print(is_subtree(tree.root, tree.root.left_child))


if __name__ == '__main__':
    main()
